climada_petals/entity/exposures/openstreetmap/osm_dataloader.py

```diff
@@ -0,0 +1,945 @@
+        -------
+        query : str
+            an SQL query string.
+        """
+        # columns which to report in output
+        query =  "SELECT osm_id"
+        for key in constraint_dict['osm_keys']:
+            query+= ","+ key
+        # filter condition(s)
+        query+= " FROM " + geo_type + " WHERE " + constraint_dict['osm_query']
+
+        return query
+
+    def retrieve(self, geo_type, constraint_dict):
+        """
+        Function to extract geometries and tag info for entires in the OSM file
+        matching certain OSM key-value constraints.
+        from an OpenStreetMap osm.pbf file.
+        adapted from BenDickens/trails repo
+        (https://github.com/BenDickens/trails.git, see extract.py)
+        
+
+        Parameters
+        ----------
+        geo_type : str
+            Type of geometry to retrieve. One of [points, lines, multipolygons]
+        constraint_dict :  dict
+            A dict with the keys "osm_keys" and "osm_query". osm_keys contains
+            a list with all the osm keys that should be reported as columns in
+            the output gdf. 
+            osm_query contains an osm query string of the syntax
+            "key(='value') (and/or further queries)".
+            See examples in DICT_CIS_OSM in case of doubt.
+
+        Returns
+        -------
+        gpd.GeoDataFrame
+            A gdf with all results from the osm.pbf file matching the 
+            specified constraints.
+
+        Note
+        ----
+        1) The keys that are searchable are specified in the osmconf.ini file.
+        Make sure that they exist in the attributes=... paragraph under the 
+        respective geometry section. 
+        For example, to retrieve multipolygons with building='yes', 
+        building must be in the attributes under
+        the [multipolygons] section of the file. You can find it in the same 
+        folder as the osm_dataloader.py module is located.
+        2) OSM keys that have : in their name must be changed to _ in the 
+                search dict, but not in the osmconf.ini
+        E.g. tower:type is called tower_type, since it would interfere with the
+        SQL syntax otherwise, but still tower:type in the osmconf.ini
+        
+        See also
+        --------
+        https://taginfo.openstreetmap.org/ to check what keys and key/value
+        pairs are valid.
+        https://overpass-turbo.eu/ for a direct visual output of the query,
+        and to quickly check the validity. The wizard can help you find the
+        correct keys / values you are looking for.
+        """
+        driver = ogr.GetDriverByName('OSM')
+        data = driver.Open(self.osm_path)
+        query = self._query_builder(geo_type, constraint_dict)
+        sql_lyr = data.ExecuteSQL(query)
+        features = []
+        if data is not None:
+            LOGGER.info('query is finished, lets start the loop')
+            for feature in tqdm(sql_lyr, desc=f'extract {geo_type}'):
```

Comment:
this is selfish, but i'd like this to run faster. could you vectorise it?
or log it as an issue for someone to solve later

you might be noticing that i'm a fan of saving nice-but-not-necessary changes as issues

Reply:
You can make it an issue 😄   I actually find it pretty fast (given the other methods of API extraction). But yeah, absolutely no clue how to set up sql queries, and even less so on how to vectorize it. This was really complicated already.

Reply:
i wasn't thinking anything fancy - probably converting the `for` loop into a list comprehension by putting the contents into a function. if i need to (try to) speed it up i'll do it :)